utils/src/main/java/software/amazon/awssdk/utils/cache/CachedSupplier.java

```diff
@@ -156,54 +188,117 @@ private void refreshCache() {
+            throw new IllegalStateException("Interrupted waiting to refresh a cached value.", e);
         }
     }
 
-    private void handleInterruptedException(String message, InterruptedException cause) {
-        Thread.currentThread().interrupt();
-        throw new IllegalStateException(message, cause);
+    /**
+     * Perform necessary transformations of the successfully-fetched value based on the stale value behavior of this supplier.
+     */
+    private RefreshResult<T> handleFetchedSuccess(RefreshResult<T> fetch) {
+        consecutiveStaleRetrievalFailures.set(0);
+
+        Instant now = clock.instant();
+
+        if (now.isBefore(fetch.staleTime())) {
+            return fetch;
+        }
+
+        switch (staleValueBehavior) {
+            case STRICT:
+                Instant newStale = now.plusSeconds(1);
+                log.warn(() -> "Retrieved value expiration is in the past (" + fetch.staleTime() + "). Using expiration "
+                               + "of " + newStale);
+                return fetch.toBuilder().staleTime(newStale).build(); // Refresh again in 1 second
+            case ALLOW:
+                Instant newStaleTime = jitterTime(now, Duration.ofMinutes(1), Duration.ofMinutes(10));
+                log.warn(() -> "Cached value expiration has been extended to " + newStaleTime + " because the downstream "
+                               + "service returned a time in the past: " + fetch.staleTime());
+
+                return fetch.toBuilder()
+                            .staleTime(newStaleTime)
+                            .build();
+            default:
+                throw new IllegalStateException("Unknown stale-value-behavior: " + staleValueBehavior);
+        }
+    }
+
+    /**
+     * Perform necessary transformations of the currently-cached value based on the stale value behavior of this supplier.
+     */
+    private RefreshResult<T> handleFetchFailure(RuntimeException e) {
+        RefreshResult<T> currentCachedValue = cachedValue;
+        if (currentCachedValue == null) {
+            throw e;
+        }
+
+        Instant now = clock.instant();
+        if (!now.isBefore(currentCachedValue.staleTime())) {
+            int numFailures = consecutiveStaleRetrievalFailures.incrementAndGet();
+
+            switch (staleValueBehavior) {
+                case STRICT:
+                    throw e;
+                case ALLOW:
+                    Instant newStaleTime = jitterTime(now, Duration.ofMillis(1), maxStaleFailureJitter(numFailures));
+                    log.warn(() -> "Cached value expiration has been extended to " + newStaleTime + " because calling the "
+                                   + "downstream service failed (consecutive failures: " + numFailures + ").");
+
+                    return currentCachedValue.toBuilder()
+                                             .staleTime(newStaleTime)
+                                             .build();
```

Comment:
Static stability! Kind of! Technically, static stability for IMDS was only supposed to account for when IMDS returned expired credentials, not when IMDS was down and we had a stale value. I took some liberties with this logic, to align with what I think would be "correct" for a cached supplier to do in this circumstance: return the stale cached value, but schedule the next 'stale' refresh to happen using exponential backoff with jitter (here, first failure: 1-100ms, second failure: 1-200ms, third failure: 1-400ms, ... up to 1-10,000 ms).